pkg/reconciler/workspace/controller.go

```diff
@@ -286,39 +288,105 @@ func (c *Controller) process(ctx context.Context, key string) error {
+				workspace.Status.Location.Current = ""
+				workspace.Status.BaseURL = ""
+			}
 		}
-		shard = currentShard
-	}
-	if workspace.Status.Location.Current == "" {
-		// find a shard for this workspace, randomly
-		shards, err := c.rootWorkspaceShardLister.List(labels.Everything())
-		if err != nil {
-			return err
+
+		if workspace.Status.Location.Current == "" {
+			// find a shard for this workspace, randomly
+			shards, err := c.rootWorkspaceShardLister.List(labels.Everything())
+			if err != nil {
+				return err
+			}
+
+			validShards := make([]*tenancyv1alpha1.WorkspaceShard, 0, len(shards))
+			invalidShards := map[string]struct {
+				reason, message string
+			}{}
+			for _, shard := range shards {
+				if valid, reason, message := isValidShard(shard); valid {
+					validShards = append(validShards, shard)
+				} else {
+					invalidShards[shard.Name] = struct {
+						reason, message string
+					}{
+						reason:  reason,
+						message: message,
+					}
+				}
+			}
+
+			if len(validShards) > 0 {
+				targetShard := shards[rand.Intn(len(shards))]
+
+				u, err := url.Parse(targetShard.Status.ConnectionInfo.Host)
+				if err != nil {
+					// shouldn't happen since we just checked in isValidShard
+					conditions.MarkFalse(workspace, tenancyv1alpha1.WorkspaceScheduled, tenancyv1alpha1.WorkspaceReasonReasonUnknown, conditionsv1alpha1.ConditionSeverityError, "Invalid connection information on target WorkspaceShard: %v.", err)
+					return err // requeue
+				}
+				logicalCluster, err := tenancyhelper.EncodeLogicalClusterName(workspace)
+				if err != nil {
+					// shouldn't happen since clusterName is supposed to be a valid name
+					conditions.MarkFalse(workspace, tenancyv1alpha1.WorkspaceScheduled, tenancyv1alpha1.WorkspaceReasonReasonUnknown, conditionsv1alpha1.ConditionSeverityError, "Invalid ClusterWorkspace cluster name %q: %v.", workspace.ClusterName, err)
+					return nil // no hope requeue fixes it
+				}
+				u.Path = path.Join(u.Path, targetShard.Status.ConnectionInfo.APIPath, "clusters", logicalCluster)
+
+				workspace.Status.BaseURL = u.String()
+				workspace.Status.Location.Current = targetShard.Name
+
+				conditions.MarkTrue(workspace, tenancyv1alpha1.WorkspaceScheduled)
+				klog.Infof("Scheduled workspace %q|%q to %q|%q", workspace.ClusterName, workspace.Name, targetShard.ClusterName, targetShard.Name)
+			} else {
+				conditions.MarkFalse(workspace, tenancyv1alpha1.WorkspaceScheduled, tenancyv1alpha1.WorkspaceReasonUnschedulable, conditionsv1alpha1.ConditionSeverityError, "No available shards to schedule workspace %q|%q.", workspace.ClusterName, workspace.Name)
+				failures := make([]string, 0, len(invalidShards))
+				for name, x := range invalidShards {
+					failures = append(failures, fmt.Sprintf("  %s: reason %q, message %q", name, x.reason, x.message))
+				}
+				klog.Infof("No valid shards found for workspace %q|%q, skipped:\n%s", workspace.ClusterName, workspace.Name, strings.Join(failures, "\n"))
```

Comment:
Ditto - can we omit workspace clusterName and name from the message?

Reply:
this is for logging, not exposed